packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,159 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
+    for (final MapEntry<String, List<Plugin>> implCandidatesEntry in pluginImplCandidates.entries) {
+      try {
+        final Plugin? resolution = _resolveImplementationOfPlugin(
+          platformKey,
+          implCandidatesEntry,
+          defaultPackageName: defaultImplementations[implCandidatesEntry.key],
+        );
+        if (resolution != null) {
+          pluginResolution[implCandidatesEntry.key] = resolution;
         }
-        continue;
-      }
-      // Next, defer to the default implementation if there is one.
-      final String? defaultPackageName = defaultImplementations[entry.key];
-      if (defaultPackageName != null) {
-        final int defaultIndex = candidates
-            .indexWhere((Plugin plugin) => plugin.name == defaultPackageName);
-        if (defaultIndex != -1) {
-          pluginResolution.add(candidates[defaultIndex]);
-          continue;
+      } on ToolExit catch (e) {
+        if (e.message != null) {
+          globals.printError(e.message!);
         }
-      }
-      // Otherwise, require an explicit choice.
-      if (candidates.length > 1) {
-        globals.printError(
-            'Plugin ${entry.key}:$platformKey has multiple possible implementations:\n'
-                '${candidates.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
-        );
         hasResolutionError = true;
-        continue;
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+/// Get the [resolutionPluginName] as the first choice for the plugin
+/// implementation and the [defaultImplementation] as fallback.
+(String? resolutionPluginName, String? defaultImplementation) _getPluginImplementationCandidate(Plugin plugin, String platformKey) {
+  String? defaultImplementation = plugin.defaultPackagePlatforms[platformKey];
+  if (plugin.platforms[platformKey] == null && defaultImplementation == null) {
+    // The plugin doesn't implement this platform.
+    return (null, null);
+  }
+  String? implementsPackage = plugin.implementsPackage;
+  final bool hasInlineDartImplementation =
+      plugin.pluginDartClassPlatforms[platformKey] != null &&
+          plugin.pluginDartClassPlatforms[platformKey] != 'none';
+  if (implementsPackage == null || implementsPackage.isEmpty) {
+    if (defaultImplementation == null && !hasInlineDartImplementation) {
+      // Skip native inline PluginPlatform implementation
+      return (null, null);
+    }
+    if (defaultImplementation != null) {
+      return (null, defaultImplementation);
```

Comment:
It is unclear to me what this means relative to the description of what the method returns. How can there be a default implementation, but no first choice implementation?

Reply:
Done in 12a38f9921d0390979945fcfe7e04f03842f2246.